# In-Class Coding Lab: Files

The goals of this lab are to help you to understand:

- Reading data from a file all at once or one line at a time.
- Searching for data in files
- Parsing text data to numerical data.
- How to build complex programs incrementally.

## Average  Spam Confidence

For this lab, we will write a program to read spam confidence headers from a mailbox file like `ICCL-mbox-tiny.txt` or `ICCL-mbox-small.txt`. These files contain raw email data, and in that data is a SPAM confidence number for each message:

`X-DSPAM-Confidence:0.8475`

Our goal will be to find each of these lines in the file, and extract the confidence number (In this case `0.8475`), with the end-goal of calculating the average SPAM Confidence of all the emails in the file. 

### Reading from the file

Let's start with some code to read the lines of text from `ICCL-mbox-tiny.txt`

In [11]:
lines = 0
filename = "CCL-mbox-tiny.txt"
with open(filename, 'r') as f:
    for line in f.readlines():
        lines = lines + 1
        #print(line.strip())        
print(lines)

332


### Now Try It

Now modify the code above to print the number of lines in the file, instead of printing the lines themselves. You'll need to increment a variable each time through the loop and then print it out afterwards.

There should be **332** lines.

### Finding the SPAM Confidence

Next, we'll focus on only getting lines addressing `X-DSPAM-Confidence:`. We do this by including an `if` statement inside  the `for` loop.

You need to edit line 4 of the code below to only print lines which begin with `X-DSPAM-Confidence:` There should be **5**

In [12]:
filename = "ICCL-mbox-tiny.txt"
with open(filename, 'r') as f:
    for line in f.readlines():
        if line.strip("X-DSPAM-Confidence:"): 
            print(line.strip())        


FileNotFoundError: [Errno 2] No such file or directory: 'ICCL-mbox-tiny.txt'

### Parsing out the confidence value

The final step is to figure out how to parse out the confidence value from the string. 
For example for the given line: `X-DSPAM-Confidence: 0.8475` we need to get the value `0.8475` as a float.

The  strategy here is to replace `X-DSPAM-Confidence:` with an empty string, then calling the `float()` function to convert the results to a float. 

### Now Try It


In [14]:
line = 'X-DSPAM-Confidence: 0.8475'
number =  line[21:len(line)]#TODO remove 'X-DSPAM-Confidence:' , then convert to a float.
number = float(number)
print (number)

0.8475


### Putting it all together

Now that we have all the working parts, let's put it all together.

```
1.  line count is 0
2.  total confidence is 0
3.  open mailbox file
4.  for each line in file
5.  if line starts with `X-DSPAM-Confidence:`
6.     remove `X-DSPAM-Confidence:` from line and convert to float
7.     increment line count
8.     add spam confidence to total confidence
9.  print average confidence (total confidence/line count)
```

In [ ]:
## TODO: Write program here:
line = 0
total = 0
filename = "ICCL-mbox-tiny.txt"
with open(filename, 'r') as f:
    for line in f.readlines():
        if line.strip("X-DSPAM-Confidence:"):
            spam =  line[21:len(line)]
            spam = float(spam)
            line = line + 1
            total = total + spam
            print(average = total / )

When you have the program working, try it with `ICCL-mbox-short.txt` mailbox file, too.